Client Support for Docker Registry HTTP API V2

stevvooe opened the pull request:
This PR updates Docker core to support pushes and pulls of images from registries that support Docker Registry HTTP API V2, as proposed in docker/docker#9015. Specifically, functionality has been added to support the V2 API while minimizing changes to the end user experience. The reasons behind this move are to improve security and access control by moving to content addressable layers and using sign-able manifests.

This PR affects the following behavior in docker core:
- `docker login`
- Pull of images.
- Push of images

While the distribution side of docker core may need some refactoring work, the goal of this particular PR is to provide equivalent functionality with the new registry API. Use of a V2 registry should be transparent to the user.

This set of changes will be supported by some accompanying PR:
- Docker Registry HTTP V2 specification (unsubmitted, based on docker/docker#9015).
- docker logout (unsubmitted)
## Overview

Images may now be pushed and pulled from registries, public and private, that implement the Docker Registry HTTP API V2.  Using `registry/v2.URLBuilder`, the client code has been ported to use the URI layout described in the proposal. Definitions for routes and error codes from [next generation docker registry](https://github.com/docker/docker-registry/tree/next-generation) are included. The rest of the changes are devoted to authentication handling and modifying the flow to support image push. 

During login, if a registry supports the V2 api, the client will use the base endpoint (“/v2/”) to solicit an authentication challenge. The challenge is then used to login to the appropriate authentication provider. The results are tested against the base endpoint to confirm the credentials are correct. After login, authorization of operations on the registry is maintained using challenges and the stored credentials. If a challenge is issued during an operation against the registry, the client attempts to re-authorize the operation, if possible.

Use of the new image format, covered in docker/docker#8093, is effectively transparent. Image manifests are created and signed transparently. This paves the way for future image verification features. A new API endpoint has been added to the docker engine to get the image manifest for an image. This endpoint can be used to get the image manifest from a daemon, sign it on the client, and push the signed image. The push API endpoint has also been updated to allow a signed manifest to be sent in the post body.

Pulling images is roughly the same as before, with most modifications to this code path consisting of URL updates and authentication management. For pushing images, the POST + PUT monolithic flow is implemented as proposed.
## Testing

There are numerous cases under which the new registry may be used. We suggest that reviewers do take the time to test cases that they care about directly. To ensure that this PR is somewhat ready, we have tested the following combinations against the [next generation registry](https://github.com/docker/docker-registry/tree/next-generation):
- “Production” registry with token auth
- Private registry with token auth
- Private registry with basic auth
- Private registry with no auth

If you need help with reproducing these scenarios, please contact us directly.

dmcgowan committed:
Push flow

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Use tarsum dev version to fix mtime issue

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Update manifest format for push

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Add trust key creation on client

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Update push to use mount blob endpoint

Using mount blob prevents repushing images which have already been uploaded

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

Commit:
Adds support for v2 registry login

summary of changes:

registry/auth.go
  - More logging around the login functions
  - split Login() out to handle different code paths for v1 (unchanged logic)
    and v2 (does not currently do account creation)
  - handling for either basic or token based login attempts
registry/authchallenge.go
  - New File
  - credit to Brian Bland <brian.bland@docker.com> (github: BrianBland)
  - handles parsing of WWW-Authenticate response headers
registry/endpoint.go
  - EVEN MOAR LOGGING
  - Many edits throught to make the coad less dense. Sparse code is more
    readable code.
  - slit Ping() out to handle different code paths for v1 (unchanged logic)
    and v2.
  - Updated Endpoint struct type to include an entry for authorization
    challenges discovered during ping of a v2 registry.
  - If registry endpoint version is unknown, v2 code path is first attempted,
    then fallback to v1 upon failure.
registry/service.go
  - STILL MOAR LOGGING
  - simplified the logic around starting the 'auth' job.
registry/session.go
  - updated use of a registry.Endpoint struct field.
registry/token.go
  - New File
  - Handles getting token from the parameters of a token auth challenge.
  - Modified from function written by Brian Bland (see above credit).
registry/types.go
  - Removed 'DefaultAPIVersion' in lieu of 'APIVersionUnknown = 0'`

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

stevvooe committed:
Registry V2 HTTP route and error code definitions

This package, ported from next-generation docker regsitry, includes route and
error definitions. These facilitate compliant V2 client implementation. The
portions of the HTTP API that are included in this package are considered to be
locked down and should only be changed through a careful change proposal.
Descriptor definitions package layout may change without affecting API behavior
until the exported Go API is ready to be locked down.

When the new registry stabilizes and becomes the master branch, this package
can be vendored from the registry.

Signed-off-by: Stephen J Day <stephen.day@docker.com>

Commit:
Update token response handling

Registry authorization token is now taken from the response body rather than
the repsonse header.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

stevvooe committed:
Remove dependencies on registry packages

Because docker core cannot vendor non-master Go dependencies, we need to remove
dependencies on registry package. The definition of digest.Digest has been
changed to a string and the regular expressions have been ported from
docker-registry/common library.

We'll likely change this be dependent on the registry in the future when the
API stabilizies and use of the master branch becomes the norm.

Signed-off-by: Stephen J Day <stephen.day@docker.com>

dmcgowan committed:
Update push and pull to registry 2.1 specification

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Allow private V2 registry endpoints

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Get token on each request

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

stevvooe committed:
Correctly check and propagate errors in v2 session

Signed-off-by: Stephen J Day <stephen.day@docker.com>

dmcgowan committed:
Fix tests

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

Commit:
Add Tarsum Calculation during v2 Pull operation

While the v2 pull operation is writing the body of the layer blob to disk
it now computes the tarsum checksum of the archive before extracting it to
the backend storage driver. If the checksum does not match that from the
image manifest an error is raised.

Also adds more debug logging to the pull operation and fixes existing test
cases which were failing. Adds a reverse lookup constructor to the tarsum
package so that you can get a tarsum object using a checksum label.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

stevvooe commented on registry/session_v2.go:
```diff
@@ -267,80 +192,82 @@ func (r *Session) GetV2ImageBlobReader(imageName, sumType, sum string, token []s
 	}
-	setTokenAuth(req, token)
-	res, _, err := r.doRequest(req)
+	queryParams := url.Values{}
```
These parameter additions need to be additive. With this approach, any query parameters in the `Location` header will be destroyed. The code should be something like this:

```
query := req.URL.Query()
query.Add("digest", sumType+":"+sumStr)
req.URL.RawQuery = query.Encode()
```

dmcgowan committed:
Refactor from feedback

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Update push to sign with the daemon's key when no manifest is given

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

LK4D4 commented:
```
docker pull -a registry-staging.docker.io/lk4d4/ipv6
Pulling repository registry-staging.docker.io/lk4d4/ipv6
FATA[0001] HTTP code: 405
```
pulling without `-a`
Also 405 pretty weird code to see :)

LK4D4 commented:
also seems 405 everywhere, on nonexistent images too.

stevvooe commented:
@LK4D4 What are the corresponding daemon logs? A 405 is returned when no handlers are matched by a particular method (which is an implementation detail of gorilla/mux).

jfrazelle commented:
theres an empty log somewhere:
```
DEBU[0313] () does not match latest (id: 7ee8afe2672221108b47b95a284eefb22cf37cd69689f7720c9b78baa9d3796e), skipping 
DEBU[0313] () does not match latest (id: d86c3be7a1464c543442f8f8ae927317cf264d34523c33c33320620a8f3e152c), skipping 
DEBU[0313] () does not match latest (id: 55134ad0e5aa8251485fa051029320fb9671db410277f238c023bb810908802a), skipping 
DEBU[0313] () does not match latest (id: 2a29728ad7af2794b82bc19fcfa40ef0ec3cafb0f7c68b10006ef259acd8745a), skipping 
DEBU[0313] () does not match latest (id: 54bbc38e71e0db5e6b234058b7fcdc737fb3cea6549e9336bf28fc0bf705d14e), skipping 
DEBU[0313] () does not match latest (id: f57aea76eca6196f72c25e59b28d251c81c9493243dceea5c0636714a4443a78), skipping 
DEBU[0313] () does not match latest (id: aa9a7c60c1cccbbdc02d2cd93fb29b72d7cae68cb9669f817e1eff77436e81b3), skipping 
DEBU[0313] () does not match latest (id: 3687839bafb21be92a5b15d611c8b21d0a6fb82b1e88694bbc4057b723316384), skipping 
DEBU[0313] () does not match latest (id: cf8dc907452c970224551599da573c9e32897fc65286d942625c4c86dabd680d), skipping 
DEBU[0313] () does not match latest (id: 2b6070780f1aa4db80ff829cf6ac2c882fe9f886a6f4a8a3eb8abb3e04286977), skipping 
DEBU[0313] () does not match latest (id: 03ab93e4fba877ed4aa565d0e82963b286c422140f933a89948db3701ee38c36), skipping 
DEBU[0313] () does not match latest (id: 3a6f1deb5d2ac3dc23140769fc5510dee69110bfbea6f0812a0c75148ec0cd42), skipping 
DEBU[0313] () does not match latest (id: 39bf9508303d5d793f6cae5b2021ab68bd0954a9f588f3d734361e82204b88dc), skipping 
DEBU[0313] () does not match latest (id: 79596999b8d3f803af63e38021fc30f00218ccebb3ee1ecc802c0592caf07578), skipping 
DEBU[0313] () does not match latest (id: 13fed73e576666252c5ae0a947168404667b639c289bf3505be4cfbf899de65f), skipping 
DEBU[0313] () does not match latest (id: 80a54c830e1012cc25a31af21c56f2f165017699810cfcfcf45b4c51bc17b9f1), skipping 
DEBU[0313] () does not match latest (id: a1c03985d5a4133d463b7ed262c039c46f0daa94f3e71d6ad815a15bef0aa0f2), skipping 
DEBU[0313] () does not match latest (id: 2b750b856e28e48098dfda59427f25a6f76265479954e67467b79b84088a3122), skipping 
DEBU[0313] () does not match latest (id: 28774ca5b76efccd56f9b74b4e3f7be1ea62dc4fb48a30ce3786d946e089594d), skipping 
DEBU[0313] () does not match latest (id: e314931015bd259897b87925c7756c33235ae3a7a28eb4accf83169a5b862a5d), skipping 
DEBU[0313] () does not match latest (id: 4b30a90166db902860ee80f24681b8e5d6de11fc6ecf8a786b65929d53686d0f), skipping 
DEBU[0313] () does not match latest (id: 589faf326e3124d8ddc66c5bdc91d923fa2184b2440b2836aefd095348b5094e), skipping 
DEBU[0313] () does not match latest (id: e267bd800f9feda4564db7583bf0aeb7032c3af6f70008d9329f01d5753946f3), skipping 
DEBU[0313] () does not match latest (id: 60bebdd2ecf9683851859d705d1df77db352e502300d9742825dd7dca483a47b), skipping 
DEBU[0313] () does not match latest (id: 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158), skipping 
DEBU[0313] () does not match latest (id: 78096d7feb0cfc711454084d102bc47dba4537ee94065cdadce44d30117ce228), skipping 
DEBU[0313] () does not match latest (id: c5881f11ded97fd2252adf93268114329e985624c5d7bb86e439a36109d1124e), skipping 
DEBU[0313] () does not match latest (id: 8955ec51479c1b4f8014a99321e4eef3283f2c0e77b5d20708eb8e74d49bf126), skipping 
DEBU[0313] () does not match latest (id: 4daf60352bebcd87df6a0eb5a66ed1921b84dde0150d529cd492eb37946b6e01), skipping 
DEBU[0313] () does not match latest (id: 4ecde1041742d1199fa54b99fcbb27f5fb49d142725ac7d9ec04ba3b5e725bbb), skipping 
DEBU[0313] () does not match latest (id: a37d6f68bd72978d5bc78d18af1b4e1380b3b811697713633fb2ff9d03a101ef), skipping 
DEBU[0313] () does not match latest (id: 707b6b3832943170244d6882511d3bfdf6f767b4c3a7b7d4dcb5c33d8fdb6f76), skipping 
DEBU[0313] () does not match latest (id: 4b908a3535f1b1a56183f1696917b02ae14f18bd76794c66d81ea1da84c57402), skipping 
DEBU[0313] () does not match latest (id: bf666b035f5c27ea69995a0e749213f2457b96480776bc059e35ebe5b9a5d44f), skipping 
DEBU[0313] () does not match latest (id: 88b37d759e1f8b1714d62e1bd5d8504e09f177750d08f18040294859104cefbd), skipping 
DEBU[0313] () does not match latest (id: fff99b7334efdac07be91f5bceb86606797e0e1393814599d3fdd39b83403c7c), skipping 
DEBU[0313] () does not match latest (id: 26bf7cebb55e49641b97910ceaa1fe766a2749ea1f7bf83c71e96c606c74abb5), skipping 
DEBU[0313] () does not match latest (id: f6168aa9c50946507acd4c874d85d454ffe573e9d054bc062ed02cb357426ad8), skipping 
......
```
what should be in the `()`

jfrazelle commented:
who is this log helping w the byte output?
```
DEBU[0385] [tar autodetect] n: [31 139 8 0 0 9 110 136 0 255]        
DEBU[0385] [tar autodetect] n: [118 97 114 47 108 105 98 47 100 112] 
```

jfrazelle commented:
```
$ docker push jess/mutt
511136ea3c5a: Image push failed 
FATA[0012] Put ?digest=tarsum.v1%2Bsha256%3Ae3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855: unsupported protocol scheme "" 
```

dmcgowan committed:
Fix list tags

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
Fixed `pull -a` in last commit.

@jfrazelle what modifications were made to support `docker push jess/mutt`, does it work when explicitly specifying a v2 registry?

We need to figure out what the proper error handling will be when attempting to fallback to v1 on a v2 only private registry.  This is what is causing the 405.

jfrazelle commented:
how do i explicitly specify as v2?

LK4D4 commented:
Also drone failed

LK4D4 commented:
Okay, because this will be used in production since `1.5` we definitely need tests. So at monday I and @icecrime will try to up local registry v2 in our image and try to write couple of tests and then you'll try to rebase and write as much tests as you can.

LK4D4 committed:
Install registry V2 in image

Signed-off-by: Alexander Morozov <lk4d4@docker.com>

LK4D4 committed:
RegistryV2 datastructure for tests

Signed-off-by: Alexander Morozov <lk4d4@docker.com>

dmcgowan committed:
Fix integration test failures

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

LK4D4 committed:
Tests for push to registry v2

Signed-off-by: Alexander Morozov <lk4d4@docker.com>

BrianBland commented:
I found an issue with the progress reporting for push with this change which can be quite misleading.

It appears that the layers are pushed in parallel (which is awesome), but if a large layer (call it A) is followed by a small layer (B) in sequential order, there is a very large chance that layer B will print an "Image already exists" message.

Looking at the logs from the registry side, it is clear that B is actually pushed, but from the engine/CLI side of things, at the time that layer A is finished being pushed and the client progress printer gets to B, the engine actually checks the registry for the remote existence of B (which does now exists). Now, instead of displaying an "Image successfully pushed" message, we get the aforementioned misleading "Image already exists".

Certainly this would be fixed by displaying each of the progress bars at the same time, but in the interim we should at least changed this message.

Commit:
Add some push test coverage

Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>

jfrazelle commented:
I'm going to swap my binary out with this and see what I run into in the next day

icecrime commented on graph/manifest.go:
```diff
@@ -0,0 +1,196 @@
+	// Get the top-most layer id which the tag points to
+	layerId, exists := localRepo[tag]
+	if !exists {
+		return nil, fmt.Errorf("Tag does not exist for %s: %s", localName, tag)
```
Nitpick: I just noticed an extra space between name and tag

Commit:
Test pulling image with aliases

Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>

dmcgowan committed:
Refactor push and pull to move code out of cmd function

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

icecrime commented on registry/auth.go:
```diff
@@ -36,6 +37,73 @@ type ConfigFile struct {
+	}
+}
+
+func (auth *RequestAuthorization) getToken() (string, error) {
```
This function has many `// TODO`, and is related to auth: what are the implications?

dmcgowan replied:
The TODOs involve optimizing when a token is retrieved.  Because the tokens are short lived, they should not be cached indefinitely.  The optimization is cache for a short period of time.  The implication of not caching is a new token is retrieved for every registry request, which will be two per layer + a call for the manifest.

dmcgowan commented:
Added some refactoring for the push and pull code which should make the diff easier to review

Also the issue mentioned by @BrianBland is not caused by parallel push (push is not parallel, only pull), rather pushing multiple empty layers.  When an empty layer is attempting to be pushed, it will see the layer as already existing since all empty layers have the same tarsum. This will appear to the user as layers already having been pushed even though the image has never been pushed.

BrianBland commented:
Yeah sorry, I got a bit confused there.

dmcgowan commented:
@LK4D4 @icecrime how can we get drone to use the correct image using the updated dockerfile?

@BrianBland it is confusing, meaning others might be confused as well.  I think the only possible way to make this particular case less confusing is by pointing out the layer is empty or skip the messaging for it entirely.

icecrime commented:
Ping @jfrazelle, our Drone feeder.

jfrazelle commented:
no we cant because drone/drone#1 and it would effect all other PRs, we can test manually. sorry.

jfrazelle commented on Dockerfile:
```diff
@@ -130,6 +130,17 @@ RUN set -x \
 	&& go install -v github.com/cpuguy83/go-md2man
 
+# Install registry
+COPY pkg/tarsum /go/src/github.com/docker/docker/pkg/tarsum
```
why?

jfrazelle replied:
shouldnt this just work? without this line?

tianon replied:
Try it, Jess. ;)

jfrazelle replied:
but but but je suis le lazyyyy, jk jk i trust you

tianon replied:
Then I'm not doing my job right.  It's always my fault, so you really
shouldn't trust me. ;)

jfrazelle commented on Dockerfile:
```diff
@@ -130,6 +130,17 @@ RUN set -x \
+	&& git clone https://github.com/docker/distribution.git /go/src/github.com/docker/distribution \
+	&& (cd /go/src/github.com/docker/distribution && git checkout -q $REGISTRY_COMMIT) \
+	&& go get -d github.com/docker/distribution/cmd/registry \
+	&& go build -o /go/bin/registry-v2 github.com/docker/distribution/cmd/registry
```
this is only for tests right or are you using this in the code?

jfrazelle replied:
you could split adding the tests into a different PR from this which might be cool

icecrime replied:
Yes, this is only for tests (we exec the `registry-v2` binary so we can `push`/`pull` to it).

TBH before this PR there was basically no tests (only 4 written, 3 of them being `t.Skip()`), so I'm not sure it's worth splitting.

jfrazelle replied:
ok no bigs just wondering

dmcgowan committed:
Cleanup v2 session to require endpoint

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Remove session backup

The v2 session code will no longer update the indexEndpoint value, therefore it is not necessary to save and restore the value for use with v1.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Add token cache

Token cache prevents the need to get a new token for every registry interaction.
Since the tokens are short lived, the cache expires after only a minute.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Fix failing integration tests

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

jfrazelle commented:
Things to note:
- Please test this PR by hand, it is only failing on drone because drone is not picking up the changes to the Dockerfile (see drone/drone#1)
- The entire test suite passes for me locally. :)
- If you would like to help test this PR and play with the binary below are links

**linux amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/linux/amd64/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/linux/amd64/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/linux/amd64/docker-1.4.1-dev.sha256

**linux 386**
http://jesss.s3.amazonaws.com/docker/pr9784/linux/386/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/linux/386/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/linux/386/docker-1.4.1-dev.sha256

**linux arm**
http://jesss.s3.amazonaws.com/docker/pr9784/linux/arm/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/linux/arm/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/linux/arm/docker-1.4.1-dev.sha256

**darwin 386**
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/386/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/386/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/386/docker-1.4.1-dev.sha256

**darwin amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/amd64/docker-1.4.1-dev
 http://jesss.s3.amazonaws.com/docker/pr9784/darwin/amd64/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/amd64/docker-1.4.1-dev.sha256

**freebsd 386**
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/386/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/386/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/386/docker-1.4.1-dev.sha256

**freebsd amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/amd64/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/amd64/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/amd64/docker-1.4.1-dev.sha256

**freebsd arm**
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/arm/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/arm/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/arm/docker-1.4.1-dev.sha256

**windows 386**
http://jesss.s3.amazonaws.com/docker/pr9784/windows/386/docker-1.4.1-dev.exe
http://jesss.s3.amazonaws.com/docker/pr9784/windows/386/docker-1.4.1-dev.exe.md5
http://jesss.s3.amazonaws.com/docker/pr9784/windows/386/docker-1.4.1-dev.exe.sha256

**windows amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/windows/amd64/docker-1.4.1-dev.exe
http://jesss.s3.amazonaws.com/docker/pr9784/windows/amd64/docker-1.4.1-dev.exe.md5
http://jesss.s3.amazonaws.com/docker/pr9784/windows/amd64/docker-1.4.1-dev.exe.sha256

jfrazelle commented:
Also just because this is such a large change, I have hooked up @dmcgowan's branch to run on jenkins here https://jenkins.dockerproject.com/job/Docker%20PR%20Test/

jfrazelle commented:
LGTM

LK4D4 commented:
LGTM
At least I tested locally :)

icecrime commented:
LGTM! @jfrazelle can we let you proceed with the merge as we have to synchronize with Drone image update?

Thanks a lot to everyone involved in reviewing this <3

stevvooe commented:
Thanks for getting this merged!
